src/main/java/org/jenkinsci/plugins/workflow/steps/ArtifactUnarchiverStepExecution.java

```diff
@@ -50,21 +51,26 @@
                 if (dst.isDirectory())
                     dst = dst.child(getFileName(all[0]));
 
-                files.add(copy(am.root().child(all[0]), dst));
+                files.add(copy(am.root().child(all[0]), dst, listener));
             } else {
                 // copy into a directory
                 for (String path : all) {
-                    files.add(copy(am.root().child(path), dst.child(path)));
+                    files.add(copy(am.root().child(path), dst.child(path), listener));
                 }
             }
         }
 
         return files;
     }
 
-    private FilePath copy(VirtualFile src, FilePath dst) throws IOException, InterruptedException {
-        try (InputStream in = src.open()) {
-            dst.copyFrom(in);
+    private FilePath copy(VirtualFile src, FilePath dst, TaskListener listener) throws IOException, InterruptedException {
+        URL u = src.toExternalURL();
+        if (u != null) {
+            new RobustHTTPClient().copyFromRemotely(dst, u, listener);
```

Comment:
Caveat that I need to look into what guarantees this is offering and how thread interruption is handled in the RobustHTTPClient